python/pyiceberg/cli/console.py

```diff
@@ -0,0 +1,404 @@
+
+    if not uri:
+        uri = os.environ.get(uri_env_var)
+    if not credential:
+        credential = os.environ.get(credential_env_var)
+
+    ctx.ensure_object(dict)
+    if output == "text":
+        ctx.obj["output"] = ConsoleOutput()
+    else:
+        ctx.obj["output"] = JsonOutput()
+
+    if not uri:
+        ctx.obj["output"].exception(
+            ValueError(f"Missing uri. Please provide using --uri or using environment variable {uri_env_var}")
+        )
+        ctx.exit(1)
+
+    assert uri  # for mypy
+
+    for scheme, catalog_type in SUPPORTED_CATALOGS.items():
+        if uri.startswith(scheme):
+            ctx.obj["catalog"] = catalog_type(catalog, uri=uri, credential=credential)  # type: ignore
+            break
+
+    if not isinstance(ctx.obj["catalog"], Catalog):
+
+        ctx.obj["output"].exception(
+            ValueError("Could not determine catalog type from uri. REST (http/https) and Hive (thrift) is supported")
+        )
+        ctx.exit(1)
+
+
+def _catalog_and_output(ctx: Context) -> Tuple[Catalog, Output]:
+    """
+    Small helper to set the types
+    """
+    return ctx.obj["catalog"], ctx.obj["output"]
+
+
+@run.command()
+@click.pass_context
+@click.argument("parent", required=False)
+def list(ctx: Context, parent: Optional[str]):  # pylint: disable=redefined-builtin
+    """Lists tables or namespaces"""
+    catalog, output = _catalog_and_output(ctx)
+
+    # still wip, will become more beautiful
+    # https://github.com/apache/iceberg/pull/5467/
+    # has been merged
+    try:
+        if parent:
+            identifiers = catalog.list_tables(parent)
+        else:
+            identifiers = catalog.list_namespaces()
+        output.identifiers(identifiers)
+    except Exception as exc:
+        output.exception(exc)
+        ctx.exit(1)
+
+
+@run.command()
+@click.option("--entity", type=click.Choice(["any", "namespace", "table"]), default="any")
+@click.argument("identifier")
+@click.pass_context
+def describe(ctx: Context, entity: Literal["name", "namespace", "table"], identifier: str):
+    """Describes a namespace xor table"""
+    catalog, output = _catalog_and_output(ctx)
+    identifier_tuple = Catalog.identifier_to_tuple(identifier)
+
+    if entity in {"namespace", "any"} and len(identifier_tuple) > 0:
+        try:
+            namespace_properties = catalog.load_namespace_properties(identifier_tuple)
+            output.describe_properties(namespace_properties)
+            ctx.exit(0)
+        except NoSuchNamespaceError as exc:
+            if entity != "any":
+                output.exception(exc)
+                ctx.exit(1)
+
+    if entity in {"table", "any"} and len(identifier_tuple) > 1:
+        try:
+            catalog_table = catalog.load_table(identifier)
+            output.describe_table(catalog_table)
+            ctx.exit(0)
+        except NoSuchTableError as exc:
+            output.exception(exc)
+            ctx.exit(1)
```

Comment:
`click` doesn't have any handling for setting the return value based on thrown exceptions? It would be nice to not need all of the `exit` calls.

Reply:
It is a bit more complicated because we both output plain text and json. If we just throw an exception, the exit code will also be non-zero, but we want to format the exception based on the output. I've created a PR that will introduce a decorator that will catch the exception: https://github.com/apache/iceberg/pull/5546